pkg/templating/manifest_generator.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// yaml generator for api key attribute
+func (i Instance) generateApiKeyAttributes() string {
+	return fmt.Sprintf(`user: %s | ""`, i.formatCredentialsLocation(i.ApiKeyLabel))
+}
+
+// yaml generator for app id/ app key attributes
+func (i Instance) generateApplicationIdAttributes() string {
+	return fmt.Sprintf("properties:\n        %s\n        %s",
+		fmt.Sprintf(`%s: %s | ""`, defaultAppIdLabel, i.formatCredentialsLocation(i.AppIDLabel)),
+		fmt.Sprintf(`%s: %s | ""`, defaultAppKeyLabel, i.formatCredentialsLocation(i.AppKeyLabel)))
+}
+
+// formats an authentication attribute into appropriate string based on the credentials location
+// if no credentials location set, defaults to checking both query params and headers
+func (i Instance) formatCredentialsLocation(key string) string {
+	var formatted string
+
+	switch i.CredentialsLocation {
+	case QueryParams:
+		formatted = fmt.Sprintf(`request.query_params["%s"]`, key)
+	case Headers:
+		formatted = fmt.Sprintf(`request.headers["%s"]`, formatHeaderLabel(key))
+	default:
+		//Unspecified
+		formatted = fmt.Sprintf(`request.query_params["%s"] | request.headers["%s"]`, key, formatHeaderLabel(key))
+	}
+	return formatted
+}
+
+// GenerateListenString - creates a string from the provided Handler replacing unset/invalid values
+// with internal defaults.
+func (h Handler) GenerateListenString() string {
+	if h.ListenAddress == "" {
+		h.ListenAddress = defaultListenAddress
+	}
+
+	if h.Port == 0 {
+		h.Port = defaultListenPort
+	}
+	return fmt.Sprintf("%s:%d", h.ListenAddress, h.Port)
+}
+
+// ConditionsToMatchString returns a valid expression for Istio match condition
+func (r Rule) ConditionsToMatchString() string {
+	var matchOn string
+	for i, condition := range r.Conditions {
+		matchOn += condition
+		if i < len(r.Conditions)-1 {
+			matchOn += " && "
+		}
+	}
+	return matchOn
+}
+
+// Return a string of the provided AuthenticationMethod
+func (am AuthenticationMethod) String() string {
+	return [...]string{"Hybrid", "ApiKey", "ApplicationID"}[am]
+}
+
+// validates and parses Handler url
+func (h Handler) parseURL() (*url.URL, error) {
+	u, err := url.ParseRequestURI(h.SystemURL)
+	if err != nil {
+		return u, fmt.Errorf("error parsing provided url. ensure a valid url has been set on Handler")
+	}
+	return u, nil
+}
+
+// Generates a unique UID from the Handler struct - provided values must conform to k8 validation.
+func (h Handler) uidGenerator(url *url.URL) (string, error) {
+	var uid string
+	if url.Host == "" || h.ServiceID == "" {
+		return uid, fmt.Errorf("error generating UID. Required seeds cannot be empty")
+	}
+
+	replacer := strings.NewReplacer(":", legalCharSeparator, ".", legalCharSeparator, "/", legalCharSeparator, "--", legalCharSeparator)
+	uid = replacer.Replace(url.Host + legalCharSeparator + url.Path + legalCharSeparator + h.ServiceID)
+	// safely strip duplicate hyphens
+	uid = replacer.Replace(uid)
+
+	validationErrs := validation.IsDNS1035Label(uid)
```

Comment:
this is a bit obscure here, maybe add a comment to explain why we do this?

Reply:
Done